Key Command Panic with Bad URL

## Version of Singularity:

`3.1.0-539.g44da2646a` (`master` at time of submission)

### Expected behavior

`key pull` and `key search` should return sensible errors when passed a bad URL:

```sh
$ singularity key search -u bad://terrible blah
ERROR:   pull failed: failed to get key: unsupported protocol scheme "bad"
$ singularity key pull -u bad://terrible 8117451AD3C42C020B9C38F8BAB9DF7600CC289B
ERROR:   pull failed: failed to get key: unsupported protocol scheme "bad"
```

### Actual behavior

`key pull`, `key search` both `panic` when passed a bad URL:

```sh
Adams-MacBook-Pro:singularity adam$ singularity key search -u bad://terrible blah
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x467a260]

goroutine 1 [running]:
github.com/sylabs/singularity/pkg/sypgp.SearchPubkey(0x7ffeefbffb69, 0x4, 0x7ffeefbffb5a, 0xe, 0xc000464000, 0x234, 0xc0001c3d70, 0xc0001c3d40)
	singularity/pkg/sypgp/sypgp.go:585 +0x1c0
github.com/sylabs/singularity/cmd/internal/cli.doKeySearchCmd(...)
	singularity/cmd/internal/cli/key_search.go:44
github.com/sylabs/singularity/cmd/internal/cli.glob..func11(0x4f22ce0, 0xc0002eb260, 0x1, 0x3)
	singularity/cmd/internal/cli/key_search.go:30 +0x7a
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x4f22ce0, 0xc0000cc0f0, 0x3, 0x3, 0x4f22ce0, 0xc0000cc0f0)
	singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2ae
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x4f24960, 0x0, 0x0, 0x14)
	singularity/vendor/github.com/spf13/cobra/command.go:850 +0x2bb
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	singularity/vendor/github.com/spf13/cobra/command.go:800
github.com/sylabs/singularity/cmd/internal/cli.ExecuteSingularity()
	singularity/cmd/internal/cli/singularity.go:136 +0x2e
main.main()
	singularity/cmd/singularity/cli.go:17 +0x20
Adams-MacBook-Pro:singularity adam$ singularity key pull -u bad://terrible 8117451AD3C42C020B9C38F8BAB9DF7600CC289B
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x467aa10]

goroutine 1 [running]:
github.com/sylabs/singularity/pkg/sypgp.FetchPubkey(0x7ffeefbffb47, 0x28, 0x7ffeefbffb38, 0xe, 0xc000216000, 0x234, 0xc0001cfd00, 0x46a1501, 0xfffffffffffffffe, 0x4883fe2, ...)
	singularity/pkg/sypgp/sypgp.go:636 +0x340
github.com/sylabs/singularity/cmd/internal/cli.doKeyPullCmd(0x7ffeefbffb47, 0x28, 0x7ffeefbffb38, 0xe, 0x0, 0x0)
	singularity/cmd/internal/cli/key_pull.go:49 +0x8f
github.com/sylabs/singularity/cmd/internal/cli.glob..func9(0x4f22820, 0xc0000d1800, 0x1, 0x3)
	singularity/cmd/internal/cli/key_pull.go:33 +0x6f
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).execute(0x4f22820, 0xc0000c2090, 0x3, 0x3, 0x4f22820, 0xc0000c2090)
	singularity/vendor/github.com/spf13/cobra/command.go:766 +0x2ae
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x4f24960, 0x0, 0x0, 0x14)
	singularity/vendor/github.com/spf13/cobra/command.go:850 +0x2bb
github.com/sylabs/singularity/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	singularity/vendor/github.com/spf13/cobra/command.go:800
github.com/sylabs/singularity/cmd/internal/cli.ExecuteSingularity()
	singularity/cmd/internal/cli/singularity.go:136 +0x2e
main.main()
	singularity/cmd/singularity/cli.go:17 +0x20
```

### Steps to reproduce behavior

As above

--- a/pkg/sypgp/sypgp.go
+++ b/pkg/sypgp/sypgp.go
@@ -576,7 +576,7 @@ func SearchPubkey(search, keyserverURI, authToken string) error {
 	// Retrieve first page of search results from Key Service.
 	keyText, err := c.PKSLookup(context.TODO(), &pd, search, client.OperationIndex, true, false, nil)
 	if err != nil {
-		if err, ok := err.(*jsonresp.Error); ok && err.Code == http.StatusUnauthorized {
+		if jerr, ok := err.(*jsonresp.Error); ok && jerr.Code == http.StatusUnauthorized {
 
 			// The request failed with HTTP code unauthorized. Guide user to fix that.
 			authToken, err := helpAuthentication()
@@ -590,7 +590,7 @@ func SearchPubkey(search, keyserverURI, authToken string) error {
 			if keyText, err = c.PKSLookup(context.TODO(), &pd, search, client.OperationIndex, true, false, nil); err != nil {
 				return err
 			}
-		} else if err.Code == http.StatusNotFound {
+		} else if ok && jerr.Code == http.StatusNotFound {
 			return fmt.Errorf("no matching keys found for fingerprint")
 		} else {
 			return fmt.Errorf("failed to get key: %v", err)
@@ -628,7 +628,7 @@ func FetchPubkey(fingerprint, keyserverURI, authToken string, noPrompt bool) (op
 	// Pull key from Key Service.
 	keyText, err := c.GetKey(context.TODO(), fp)
 	if err != nil {
-		if err, ok := err.(*jsonresp.Error); ok && err.Code == http.StatusUnauthorized {
+		if jerr, ok := err.(*jsonresp.Error); ok && jerr.Code == http.StatusUnauthorized {
 
 			// The request failed with HTTP code unauthorized. Guide user to fix that.
 			authToken, err := helpAuthentication()
@@ -641,7 +641,7 @@ func FetchPubkey(fingerprint, keyserverURI, authToken string, noPrompt bool) (op
 			if keyText, err = c.GetKey(context.TODO(), fp); err != nil {
 				return nil, err
 			}
-		} else if err.Code == http.StatusNotFound {
+		} else if ok && jerr.Code == http.StatusNotFound {
 			return nil, fmt.Errorf("no matching keys found for fingerprint")
 		} else {
 			return nil, fmt.Errorf("failed to get key: %v", err)
@@ -696,7 +696,7 @@ func PushPubkey(e *openpgp.Entity, keyserverURI, authToken string) error {
 
 	// Push key to Key Service.
 	if err := c.PKSAdd(context.TODO(), keyText); err != nil {
-		if err, ok := err.(*jsonresp.Error); ok && err.Code == http.StatusUnauthorized {
+		if jerr, ok := err.(*jsonresp.Error); ok && jerr.Code == http.StatusUnauthorized {
 
 			// The request failed with HTTP code unauthorized. Guide user to fix that.
 			authToken, err := helpAuthentication()

--- a/pkg/sypgp/sypgp_test.go
+++ b/pkg/sypgp/sypgp_test.go
@@ -68,6 +68,9 @@ func TestSearchPubkey(t *testing.T) {
 	}{
 		{"Success", http.StatusOK, openpgp.EntityList{testEntity}, "search", srv.URL, "", false},
 		{"SuccessToken", http.StatusOK, openpgp.EntityList{testEntity}, "search", srv.URL, "token", false},
+		{"BadURL", http.StatusOK, openpgp.EntityList{testEntity}, "search", ":", "", true},
+		{"TerribleURL", http.StatusOK, openpgp.EntityList{testEntity}, "search", "terrible:", "", true},
+		{"NotFound", http.StatusNotFound, nil, "search", srv.URL, "", true},
 		{"Unauthorized", http.StatusUnauthorized, nil, "search", srv.URL, "", true},
 	}
 	for _, tt := range tests {
@@ -103,6 +106,7 @@ func TestFetchPubkey(t *testing.T) {
 		{"NoKeys", http.StatusOK, openpgp.EntityList{}, fp, srv.URL, "token", true},
 		{"TwoKeys", http.StatusOK, openpgp.EntityList{testEntity, testEntity}, fp, srv.URL, "token", true},
 		{"BadURL", http.StatusOK, openpgp.EntityList{testEntity}, fp, ":", "", true},
+		{"TerribleURL", http.StatusOK, openpgp.EntityList{testEntity}, fp, "terrible:", "", true},
 		{"NotFound", http.StatusNotFound, nil, fp, srv.URL, "", true},
 		{"Unauthorized", http.StatusUnauthorized, nil, fp, srv.URL, "", true},
 	}
@@ -170,20 +174,23 @@ func TestPushPubkey(t *testing.T) {
 
 	tests := []struct {
 		name      string
+		uri       string
 		authToken string
 		code      int
 		wantErr   bool
 	}{
-		{"Success", "", http.StatusOK, false},
-		{"SuccessToken", "token", http.StatusOK, false},
-		{"Unauthorized", "", http.StatusUnauthorized, true},
+		{"Success", srv.URL, "", http.StatusOK, false},
+		{"SuccessToken", srv.URL, "token", http.StatusOK, false},
+		{"BadURL", ":", "", http.StatusOK, true},
+		{"TerribleURL", "terrible:", "", http.StatusOK, true},
+		{"Unauthorized", srv.URL, "", http.StatusUnauthorized, true},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			ms.code = tt.code
 
-			if err := PushPubkey(testEntity, srv.URL, tt.authToken); (err != nil) != tt.wantErr {
+			if err := PushPubkey(testEntity, tt.uri, tt.authToken); (err != nil) != tt.wantErr {
 				t.Fatalf("got err %v, want error %v", err, tt.wantErr)
 			}
 		})

